addon/modifiers/validity.js

```diff
@@ -28,43 +34,44 @@ export default modifier(function validity(
   validators,
   { watch = [], on: eventNames = 'change,input,blur' }
 ) {
+  assert(
+    'Only one validity modifier can be applied to an element',
+    !isValidatable(element),
+  );
+  let taskCount = 0;
+  let lastTask = Promise.resolve();
   let autoValidationEvents = commaSeperate(eventNames);
-  let autoValidationHandler = () => validate(element);
-  let validateHandler = async () => {
-    let [error = ''] = await reduceValidators(validators, element);
-    element.checkValidity();
-    element.setCustomValidity(error);
-    element.dispatchEvent(new CustomEvent('validated'));
+  let updateValidity = ({ target }) => {
+    taskCount++;
+    lastTask = lastTask.then(async () => {
```

Comment:
Hmm... idk if Ember is aware of this async work and if it would properly wait for this to complete before carrying on with tests. I'm worried a test like this...

```javascript
render(hbs`
  <input {{on 'validated' this.setErrorMsg}} />
  <div data-test-error>{{this.errorMsg}}</div>
`);
await fillIn('input', 'lol this text is invalid');
assert.equal(find('[data-test-error]').innerText, 'Oops, input is invalid!');
```

wouldn't work as expected. If this doesn't work, I _think_ `ember-test-waiters` or using `RSVP` would help?

Reply:
Pretty sure since Ember supports microtasks RSVP is no longer needed. However having the test waiter might be an idea. Will muse over it.